File: src/main/java/com/team3/gudit/payment/entity/Payment.java
package com.team3.gudit.payment.entity;

import com.team3.gudit.global.exception.BusinessException;
import com.team3.gudit.payment.exception.PaymentErrorCode;
import com.team3.gudit.purchase.entity.Purchase;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(
        name = "payments",
        uniqueConstraints = {
                @UniqueConstraint(
                        name = "uk_payment_purchase",
                        columnNames = "purchase_id"
                ),
                @UniqueConstraint(
                        name = "uk_payment_order_id",
                        columnNames = "order_id"
                ),
                @UniqueConstraint(
                        name = "uk_payment_payment_key",
                        columnNames = "payment_key"
                )
        }
)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "purchase_id", nullable = false)
    private Purchase purchase;

    @Column(name = "order_id", nullable = false, length = 64)
    private String orderId;

    @Column(name = "payment_key")
    private String paymentKey;

    @Column(nullable = false)
    private int amount;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentStatus status;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "canceled_at")
    private LocalDateTime canceledAt;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    private Payment(Purchase purchase, int amount) {
        this.purchase = purchase;
        this.orderId = "GUDIT_" + UUID.randomUUID();
        this.amount = amount;
        this.status = PaymentStatus.READY;
    }

    public static Payment create(Purchase purchase, int amount) {
        return new Payment(purchase, amount);
    }

    public void start(String paymentKey) {
        validateStatus(PaymentStatus.READY);

        this.paymentKey = paymentKey;
        this.status = PaymentStatus.IN_PROGRESS;
    }

    public void complete(LocalDateTime approvedAt) {
        validateStatus(PaymentStatus.IN_PROGRESS);

        this.status = PaymentStatus.DONE;
        this.approvedAt = approvedAt;
    }

    public void fail() {
        validateStatus(PaymentStatus.IN_PROGRESS);

        this.status = PaymentStatus.FAILED;
    }

    public void cancel() {
        validateStatus(PaymentStatus.DONE);

        this.status = PaymentStatus.CANCELED;
        this.canceledAt = LocalDateTime.now();
    }

    public void cancelAfterApprovalFailure() {
        validateStatus(PaymentStatus.IN_PROGRESS);

        this.status = PaymentStatus.CANCELED;
        this.canceledAt = LocalDateTime.now();
    }

    public void cancelReady() {
        validateStatus(PaymentStatus.READY);

        this.status = PaymentStatus.CANCELED;
        this.canceledAt = LocalDateTime.now();
    }

    private void validateStatus(PaymentStatus expectedStatus) {
        if (this.status != expectedStatus) {
            throw new BusinessException(
                    PaymentErrorCode.INVALID_PAYMENT_STATUS,
                    "Invalid payment status transition. current="
                            + this.status
                            + ", expected="
                            + expectedStatus
            );
        }
    }

    public void completeByWebhook(
            String paymentKey,
            LocalDateTime approvedAt
    ) {
        if (this.status != PaymentStatus.READY
                && this.status != PaymentStatus.IN_PROGRESS) {
            throw new BusinessException(
                    PaymentErrorCode.INVALID_PAYMENT_STATUS
            );
        }

        this.paymentKey = paymentKey;
        this.status = PaymentStatus.DONE;
        this.approvedAt = approvedAt;
    }

    public void failByWebhook() {
        if (this.status != PaymentStatus.READY
                && this.status != PaymentStatus.IN_PROGRESS) {
            throw new BusinessException(
                    PaymentErrorCode.INVALID_PAYMENT_STATUS
            );
        }

        this.status = PaymentStatus.FAILED;
    }

    public void cancelByWebhook() {
        if (this.status != PaymentStatus.READY
                && this.status != PaymentStatus.IN_PROGRESS
                && this.status != PaymentStatus.DONE) {
            throw new BusinessException(
                    PaymentErrorCode.INVALID_PAYMENT_STATUS
            );
        }

        this.status = PaymentStatus.CANCELED;
        this.canceledAt = LocalDateTime.now();
    }
}

File: src/main/kotlin/com/team3/gudit/payment/entity/Payment.kt
package com.team3.gudit.payment.entity

import com.team3.gudit.global.exception.BusinessException
import com.team3.gudit.payment.exception.PaymentErrorCode
import com.team3.gudit.purchase.entity.Purchase
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.OneToOne
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import java.time.LocalDateTime
import java.util.UUID
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener

@Entity
@Table(
    name = "payments",
    uniqueConstraints = [
        UniqueConstraint(name = "uk_payment_purchase", columnNames = ["purchase_id"]),
        UniqueConstraint(name = "uk_payment_order_id", columnNames = ["order_id"]),
        UniqueConstraint(name = "uk_payment_payment_key", columnNames = ["payment_key"])
    ]
)
@EntityListeners(AuditingEntityListener::class)
class Payment private constructor(purchase: Purchase, amount: Int) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
        protected set

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "purchase_id", nullable = false)
    var purchase: Purchase = purchase
        protected set

    @Column(name = "order_id", nullable = false, length = 64)
    var orderId: String = "GUDIT_${UUID.randomUUID()}"
        protected set

    @Column(name = "payment_key")
    var paymentKey: String? = null
        protected set

    @Column(nullable = false)
    var amount: Int = amount
        protected set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: PaymentStatus = PaymentStatus.READY
        protected set

    @Column(name = "approved_at")
    var approvedAt: LocalDateTime? = null
        protected set

    @Column(name = "canceled_at")
    var canceledAt: LocalDateTime? = null
        protected set

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: LocalDateTime? = null
        protected set

    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime? = null
        protected set

    fun start(paymentKey: String?) {
        validateStatus(PaymentStatus.READY)
        this.paymentKey = paymentKey
        status = PaymentStatus.IN_PROGRESS
    }

    fun complete(approvedAt: LocalDateTime) {
        validateStatus(PaymentStatus.IN_PROGRESS)
        status = PaymentStatus.DONE
        this.approvedAt = approvedAt
    }

    fun fail() {
        validateStatus(PaymentStatus.IN_PROGRESS)
        status = PaymentStatus.FAILED
    }

    fun cancel() {
        validateStatus(PaymentStatus.DONE)
        status = PaymentStatus.CANCELED
        canceledAt = LocalDateTime.now()
    }

    fun cancelAfterApprovalFailure() {
        validateStatus(PaymentStatus.IN_PROGRESS)
        status = PaymentStatus.CANCELED
        canceledAt = LocalDateTime.now()
    }

    fun cancelReady() {
        validateStatus(PaymentStatus.READY)
        status = PaymentStatus.CANCELED
        canceledAt = LocalDateTime.now()
    }

    private fun validateStatus(expectedStatus: PaymentStatus) {
        if (status != expectedStatus) {
            throw BusinessException(
                PaymentErrorCode.INVALID_PAYMENT_STATUS,
                "Invalid payment status transition. current=$status, expected=$expectedStatus"
            )
        }
    }

    fun completeByWebhook(paymentKey: String?, approvedAt: LocalDateTime) {
        if (status != PaymentStatus.READY && status != PaymentStatus.IN_PROGRESS) {
            throw BusinessException(PaymentErrorCode.INVALID_PAYMENT_STATUS)
        }
        this.paymentKey = paymentKey
        status = PaymentStatus.DONE
        this.approvedAt = approvedAt
    }

    fun failByWebhook() {
        if (status != PaymentStatus.READY && status != PaymentStatus.IN_PROGRESS) {
            throw BusinessException(PaymentErrorCode.INVALID_PAYMENT_STATUS)
        }
        status = PaymentStatus.FAILED
    }

    fun cancelByWebhook() {
        if (status != PaymentStatus.READY &&
            status != PaymentStatus.IN_PROGRESS &&
            status != PaymentStatus.DONE
        ) {
            throw BusinessException(PaymentErrorCode.INVALID_PAYMENT_STATUS)
        }
        status = PaymentStatus.CANCELED
        canceledAt = LocalDateTime.now()
    }

    companion object {
        @JvmStatic
        fun create(purchase: Purchase, amount: Int): Payment = Payment(purchase, amount)
    }
}
